Framework/PythonInterface/plugins/algorithms/LinkedUBs.py

```diff
@@ -0,0 +1,444 @@
+                d_predicted.append(_d_predicted)
+                ind_h.append(_h_predicted)
+                ind_k.append(_k_predicted)
+                ind_l.append(_l_predicted)
+
+            # recast these lists as numpy arrays
+            q_predicted = np.array(q_predicted)
+            d_predicted = np.array(d_predicted)
+
+            ind_h = np.array(ind_h)
+            ind_k = np.array(ind_k)
+            ind_l = np.array(ind_l)
+
+            # get the indexing list that sorts dspacing from largest to
+            # smallest
+            idx = d_predicted.argsort()[::-1]
+
+            # sort q, d and h, k, l by this indexing
+            q_predicted = q_predicted[idx]
+            d_predicted = d_predicted[idx]
+            ind_h = ind_h[idx]
+            ind_k = ind_k[idx]
+            ind_l = ind_l[idx]
+
+            HKL_predicted = []
+            for i in range(len(ind_h)):
+                _HKL = (ind_h[i], ind_k[i], ind_l[i])
+                HKL_predicted.append(_HKL)
+
+            # recast h,k,l list as numpy array
+            HKL_predicted = np.array(HKL_predicted)
+
+            # create the list of predicted peaks to be considered in
+            # linking procedure, with peak list partitioned by num_peaks
+            q_ordered = []
+            d_ordered = []
+            HKL_ordered = []
+
+            for i in range(0, _num_peaks_var):
+                _q_ordered = q_predicted[i]
+                _d_ordered = d_predicted[i]
+                _HKL_ordered = HKL_predicted[i]
+
+                q_ordered.append(_q_ordered)
+                d_ordered.append(_d_ordered)
+                HKL_ordered.append(_HKL_ordered)
+
+            # recast these ordered lists as numpy arrays
+            q_ordered = np.array(q_ordered)
+            d_ordered = np.array(d_ordered)
+            HKL_ordered = np.array(HKL_ordered)
+
+            # loop through the ordered find peaks, compare q and d to each
+            # predicted peak if the q and d values of a found peak match a
+            # predicted peak within tolerance, the found peak inherits
+            # the HKL of the predicted peak
+            for i in range(len(q_find)):
+                _qx_obs, _qy_obs, _qz_obs = q_find[i]
+                _q_obs = V3D(_qx_obs, _qy_obs, _qz_obs)
+                _p_obs = _LinkedPeaks.createPeak(_q_obs)
+                _d_obs = d_find[i]
+
+                for j in range(len(q_ordered)):
+                    _qx_pred, _qy_pred, _qz_pred = q_ordered[j]
+                    _d_pred = d_ordered[j]
+
+                    if (_qx_pred -
+                        _qtol_var <= _qx_obs <= _qx_pred +
+                        _qtol_var and _qy_pred -
+                        _qtol_var <= _qy_obs <= _qy_pred +
+                        _qtol_var and _qz_pred -
+                        _qtol_var <= _qz_obs <= _qz_pred +
+                        _qtol_var and _d_pred -
+                        self._dtol <= _d_obs <= _d_pred +
```

Comment:
I think the first three conditions are just checking whether the L1 norm of q observed is less than q tolerance?
So assuming `q_obs` and `q_pred` are numpy arrays we can write something like the following:

```python
if np.linalg.norm(q_obs - q_pred, ord=1) < qtol* 3.
    and np.absolute(d_pred - d_obs) < dtol:
...
```

Reply:
Thanks Sam.  

I tried various implementations of this but found that they didn't pass the system test and were generally slower. I think for now the conditions on each q component works better but open to discussion on this point. 

I agreed with and implemented the other suggestions